harrystech/arthur-redshift-etl
# Grant and revoke for groups within one statement

## Description

We used to call `GRANT` or `REVOKE` for every group individually, which is a bit expensive. Each of those calls takes up to 1s. So this PR merges the calls. Note that it's important to reduce the time it takes to cut over since during the time of promoting the staging schemas (and backup up first the original schemas), the users can sometimes not see a schema, leading to strange errors during queries.

### Example

Going from:
```
GRANT USAGE ON SCHEMA "common" TO GROUP "etl_ro"
GRANT USAGE ON SCHEMA "common" TO GROUP "analyst_ro"
```
to:
```
GRANT USAGE ON SCHEMA "common" TO GROUP "etl_ro", GROUP "analyst_ro"
```

### Other changes

In `etl.text` some utility functions are renamed to make it easier to distinguish them and use them.

Old name | New name
---|---
`join_with_quotes`  | `join_with_single_quotes`
`join_column_list` | `join_with_double_quotes`

## How to test
Schemas can be created and moved around using the schemas commands.

1. Connect to "your" validation database by copying the `credentials_validation.sh` file into your local config directory.
2. Run through the various steps
```
arthur.py create_schemas --prolix common
arthur.py create_schemas --with-backup --prolix common
arthur.py create_schemas --with-staging --prolix common
arthur.py promote_schemas --from-position staging --prolix common
arthur.py promote_schemas --from-position backup --prolix common
```
(The `--prolix` flag makes it easier to observe the queries.)

Don't forget to remove the `credentials_validation.sh` file so that you're connected to the correct database again.
